### ISSUE TITLE
Diligent Renderer integration

### ISSUE THREAD

**Issue by eugeneko** [rbfx/rbfx#547]
This issue is the aggregated task list of all the things we need to do before we can merge `diligent` to `master`.
Items are listed in order from higher to lower priority.

Pending tasks:

- [x] (!) Make CI green, whatever it takes. Every build should use Diligent as backend.
- [x] Extract Diligent thirdparty in minimal configuration.
- [x] Restore disabled components (like Editor) and implement missing aspects
- [x] Refactor and simplify rendering code
- [x] Get C# working
- [x] Revisit old defines like `GL_ES_VERSION_2_0` or `URHO3D_OPENGL`
- [x] Test and fix features, backends and platforms
- [x] Resolve all `TODO(diligent):`

@gleblebedev, @rokups, @rbnpontes you can add new items in comments so they won't get lost

**Comment by rbnpontes** [comment]
Here is some things that we can do

- Complete Shader Computer, there's no implementation. But it's simple, just get currently DeviceContext on GraphicsImpl, create compute pipeline and execute compute. 
- Fix washed rendering on OpenGL
- Fix D3D12 rendering, specifically we must make it run RenderingShowcase without problems.
- Maybe fix some issues on Mobile devices like Android and iOS
- Fix UWP, basically must compile to UWP and see if is rendering pretty well, it's easy
- Implement MSAA
- Finish Texture Resources implementation, I think I haven't finish all texture resources impl, I think that Texture3D is incomplete
- (This can be optionally, idk if it needs) making sure that diligent is running on Web builds, diligent can compile to emscripten, but I never tested
- Create unit tests ?

**Comment by eugeneko** [comment]
Given that I would like to drop all existing backends, Mobiles and Web are kinda mandatory.
I am not saying that current GL backend is *bad*, but it is quite poorly organized and outdated, and it is pulling us back.

**Comment by eugeneko** [comment]
I rebased @rbnpontes work on top of master and formatted & cleaned it up just a bit.
We are getting our first green jobs in CI. So, current `diligent` branch can be the baseline for further work.

**Comment by rbnpontes** [comment]
Great

**Comment by eugeneko** [comment]
We've got our first fully green CI build with Diligent integration. Now I can start actually working on the code.
https://github.com/rbfx/rbfx/actions/runs/4837989758

**Comment by eugeneko** [comment]
Okay, the biggest issue I see is vertex layout. I am not sure how to resolve that.

- Diligent requires input layout to be mapped via vertex element indexes. No position/normal/etc, just 0,1,2...
- Input layout should be provided during pipeline state creation,
- How can we learn layout indexes *before* pipeline state creation, especially on OpenGL?

Current prototype uses SPIRV-Reflect to peek into , but it is not viable release option because mobile/web platforms don't go thru SpirV at all, the build may not even have anything SPIR-V related.

**Comment by rbnpontes** [comment]
I don't had tested but maybe you can use macros for OpenGL and resolve into runtime.

Ex:
`#define POSITION_LOCATION 0
#define COLOR_LOCATION 1

layout(location=POSITION_LOCATION)`

On OpenGL and Vulkan environments, even though you won't use vertex input in shader, you still must to declare Input Layout list and macro approach maybe work.
This is not the same to DirectX because unused vertex elements will be removed by the shader

**Comment by eugeneko** [comment]
I think I've dealt with vertex layout, I also refactored uniforms processing a bit.
Now I am trying to get rid of fixed texture slots and just use texture names instead, because this is both something that we wanted to have in the engine and it's more Diligent-friendly.

**Comment by rbnpontes** [comment]
Great.
Change Texture Units to name based is difficult because Texture Units is heavily tied to Materials, RenderPath and Passes.

**Comment by eugeneko** [comment]
> Change Texture Units to name based is difficult because Texture Units is heavily tied to Materials, RenderPath and Passes.

I didn't change resources like Material because it's not that important right now, we can just map TextureUnit to name on the material level.
The rest is done in the latest commit.

Also, I noticed that I need to refactor SRB binding cache so it is properly invalidated when texture is e.g. reloaded. I will probably do what I did with pipeline state hashes in Material and VertexBuffer, and add some events for SRB invalidation.

**Comment by eugeneko** [comment]
@rbnpontes do you think it's worth having cached SRB, considering the issues with invalidation?
Or we can just do "Set(texture)" on every frame? Did you try this approach?

**Comment by rbnpontes** [comment]
Its not recommended because they cause performance losses 
You can make some or all resources with dynamic, in this case Diligent expects that this binding can be changed at higher frequency.

At PipelineState building you can pass what resources will be Static, Dynamic and Mutable, you can add resource after pipeline created off course.
On old backends this is not a big problem, but when you deal with modern backends like Vulkan and D3D12 this can be problem.

Usually, i use Static binding in  Render Passes, Light Passes and Shadow Passes pipelines because they won't change this resources along the frames.
The same way i recommend to use Static binding for Constant Buffers because they rarely changes.

Generally i use Mutable binding to materials. 

Pipeline State is cached, and its good but what about your resources ?
You can reuse pipeline and you can have multiple SRB's, this means that you don't need to create another pipeline just because you have changed the texture, just create a new SRB and add to.
You can add a resource to SRB even though you don't use on shader. ( I made this on DrawCommand, i bind all cbuffers in one SRB).

If you have any question about this model, you can check on the Diligent blog they explain 
 better why they did use this method [http://diligentgraphics.com/2016/03/23/resource-binding-model-in-diligent-engine-2-0/](http://diligentgraphics.com/2016/03/23/resource-binding-model-in-diligent-engine-2-0/)

**Comment by eugeneko** [comment]
> Its not recommended because they cause performance losses

Yeah, I understand that this is suboptimal. However, proper SRB invalidation is not an easy feat either. E.g we need to recreate SRB when texture filtering mode changes, or when texture is reloaded, and so on.
So, I wonder if performance loss is acceptable if we just use Mutable bindings for everything.

PS: I may have missed something, but you make all resources mutable, no? The only place where resource type is specified is the line "ci.PSODesc.ResourceLayout.DefaultVariableType = SHADER_RESOURCE_VARIABLE_TYPE_MUTABLE".
So we shouldn't get much from ditching SRB cache?

I think that SRB cache is good thing in general, but the proper implementation is much more complicated than one in your prototype. So maybe we want to postpone that until we deal with more important issues?

**Comment by rbnpontes** [comment]
Yeah  i made all resources as mutable on pipeline.

I don't see any problems, like i said on D3D11 and OpenGL maybe won't loses too much performance, idk about D3D12 and Vulkan, we need to profile.

**Comment by eugeneko** [comment]
Good news: shader processing is now like 10 times simpler.
Bad news: I broke Vulkan xD. It cannot bind sampler automatically for whatever reason.

**Comment by rbnpontes** [comment]
Do you have generated GLSL code ?

**Comment by rbnpontes** [comment]
I had compile your code, its seems the problem is at step to generate universal shader.
Your GLSL conversion to SPIRV code is not including _sampler object, for this reason this error occurs.
You can do it manually by calling `ITextureView::SetSampler` and it will work, but in this case the whole texture will share unique sampler, this lead to unexpected behaviors in some rbfx samples

**Comment by eugeneko** [comment]
Our existing backends already kinda live with shared texture samplers (one
texture - one sampler). So it shouldn’t cause anything weird, no? At least
from what I’m seeing. That’s how it always worked. We can refactor that
later, but maybe in separate task.

Ср, 10 мая 2023 г. в 00:14, rbnpontes ***@***.***>:

**Comment by rbnpontes** [comment]
Its work if you force sampler with `ITextureView::SetSampler`.

But StaticScene sample is not rendering, do you know what it could be ?
![image](https://github.com/rbfx/rbfx/assets/10354401/a3a79758-3313-4777-a738-bfc841b9fdde)
```
[19:02:09] [error] [main] : (diligent) Vulkan debug message (validation): VUID-vkCmdCopyBufferToImage-pRegions-06218
                 Validation Error: [ VUID-vkCmdCopyBufferToImage-pRegions-06218 ] Object 0: handle = 0x4e4775000000006e, 
name = Textures/Mushroom.dds, type = VK_OBJECT_TYPE_IMAGE; | MessageID = 0x9b5db4c8 | 
vkCmdCopyBufferToImage: Both pRegion[0] imageoffset.x (0) and (imageExtent.width + imageOffset.x) (4) must be >= zero or <= image subresource width (2). 
The Vulkan spec states: For each element of pRegions, imageOffset.x and (imageExtent.width + imageOffset.x) must both be greater than or equal to 0 and less than or equal to the width of the specified imageSubresource of dstImage (https://vulkan.lunarg.com/doc/view/1.2.198.1/windows/1.2-extensions/vkspec.html#VUID-vkCmdCopyBufferToImage-pRegions-06218)                 
Object[0] (image): Handle 0x4e4775000000006e, Name: 'Textures/Mushroom.dds'
```

**Comment by eugeneko** [comment]
@rbnpontes yeah, it crashes for me too, something's off with inputs heh
However, it crashes only on one of OpaquePass draws, you can still inspect others.

Apparently, the input mapping is off. Vertex shader doesn't receive the position.
I am yet to debug why exactly this happens.

Since I completely changed the way SpirV generates (because I don't want to do GLSL->SPIRV->HLSL->SPIRV for vulkan lol), this kind of issues is sadly expected.

**Comment by eugeneko** [comment]
I fixed bad input bindings on Vulkan. We still get black screen, but the depth is actually filled, so it's the issue with output now.

**Comment by rbnpontes** [comment]
Great, I will test later

**Comment by eugeneko** [comment]
@rbnpontes cbuffers were okay, but shader reflection from Diligent with Vulkan backend doesn't match its documentation heh.
I fixed that. StaticScene now works in some capacity on all 4 backends.

**Comment by rbnpontes** [comment]
Great, but what's problem with Reflection ?

**Comment by rbnpontes** [comment]
That's why diligent follow the same order of OpenGL ? I mean, on GL matrix columns and rows is not the same way of Directx.
Idk, it seems weird

**Comment by eugeneko** [comment]
I worked on using immutable samplers. I am not 100% happy with that, because we really need to carefully prepare PipelineState objects now, and sadly we create pipeline state objects in multiple places. It needs better incapsulation, maybe... Or at least warnings or even asserts when we don't have immutable sampler for the resource, so we can catch all the places.

**Comment by rbnpontes** [comment]
Why not use Immutable Samplers at first moment on Render Passes ?
Scene Passes maybe can use Dynamic samplers until we find a better way to handle this

**Comment by eugeneko** [comment]
> Scene Passes maybe can use Dynamic samplers until we find a better way to handle this

Diligent on DX12 backend cannot handle more than 2048 dynamic samplers _per frame_. Not "unique samplers", just samplers total. If you have 16 sampled textures per shader and you use dynamic samplers, then you can only change resources 128 times per frame... Which is just ridiculously low. Imagine having a limit of 50 materials per frame.

**Comment by eugeneko** [comment]
So, we either treat samplers as part of PipelineState object and process them accordingly, or we face heavy penalties.

**Comment by rbnpontes** [comment]
Agreed, i don't see any problem
Caching is here to help us, basically we can trust on PSO Cache.

The basic problem that i see is deal with unused pipelines along the frames

**Comment by eugeneko** [comment]
Diligent backend now runs RenderingShowcase warningless on all backends.
Shadows on OpenGL backend are still slightly off, but I will debug it later.

**Comment by rbnpontes** [comment]
Great. Just a reminder, please review my Compute PR. It's a small change, and I'm afraid that your changes could affect the implementation. If you have free time, of course.

I created a compute pipeline on ComputeDevice to make this feature work. I'm asking for your review because you may have a better idea of where the compute pipeline should be built.
Some of these changes, such as Samplers, are not included in Compute; they all use dynamic samplers.

**Comment by rbnpontes** [comment]
Agreed!

**Comment by rbnpontes** [comment]
I was thinking on something.
What do you think about cache Render Passes ?
I mean, we could record all render passes in json, xml or binary. Diligent had a solution for this case, but i think this will won't work very well on rbfx case.

Obviously this can go after diligent integration

**Comment by eugeneko** [comment]
I don't understand some weird issues with Web tho. In RenderingShowcase demo, there is this weird web-only bug:
- Reflective ball in reflection demo scene is not rendered.
- Probe object attached to the camera is rendered fine in main scene and in bike scene, but is not rendered in lightmap demo and reflection demo scenes.

That's really weird, yeah. I have no clue why this specific object is having issues, and why it happens only in Web.

Also, there are a couple of weird opengl errors too in Web.

**Comment by rbnpontes** [comment]
This error occurs rarely on legacy.
i had some investigations but i reached dead line without any explanations, its very weird.
If you switch the samples and go back to reflective ball sample it will work

**Comment by eugeneko** [comment]
Okay, we reached the another big milestone. The engine works with Diligent backend with green CI and without disabled components. It is still buggy and somewhere messy tho.

**Comment by eugeneko** [comment]
Update: I fixed Android (no Vulkan yet), Web and UWP.

Sadly, Vulkan on Android crashes (while Diligent samples work) and our API is too messy to bisect what exactly we are doing wrong.

**Comment by rbnpontes** [comment]
Log all graphics commands, you can rebuild all commands later and find what goes wrong

**Comment by eugeneko** [comment]
Refactoring plans:

- Continue moving stuff to **Urho3D/RenderAPI/**. We should not have anything backend-related in **Urho3D/Graphics/**, basically **Urho3D/Graphics/Diligent** should become empty. This way we will separate rendering backend from high-level graphic abstractions.
- Add sample for Texture 2D Array and Texture 3D, because they are currently not covered and therefore very hard to test
- Same goes for ComputeDevice.
- Strip as much as possible from Graphics.h/cpp, this class is really overloaded. Maybe remove old renderer at this point? It will not work with Diligent anyway.

**Comment by rbnpontes** [comment]
Why not make backend available and move this hard refactoring to another version ?

**Comment by eugeneko** [comment]
Backend still needs multiple fixes until it’s ready to ship, and I am
really not looking toward fixing old urho api just to remove it a few weeks
later. I don’t want to merge unmaintainable and/or partially broken code
and then be under the pressure to deal with it asap. Since we don’t have
releases yet, I prefer to keep “release quality” master branch.

**Comment by rbnpontes** [comment]
Sure, i fully understand.
Do you need any help ? Maybe add new samples like Compute Shader and Tessellation ?

**Comment by eugeneko** [comment]
If you have time, you can help by testing and/or patching platforms from top post (whichever are avaiable for you).
I did test what I could, but there cannot be "too much testing" heh. Also, no Mac.

Also, you can try to figure out how to enable MoltenVK in our Actions build. You can run Actions in your fork freely to test stuff.

Mac has this error but it has Vulkan enabled:
```
2023-06-05T17:38:12.5355810Z CMake Warning at Source/ThirdParty/Diligent/Graphics/GraphicsEngineVulkan/CMakeLists.txt:232 (message):
2023-06-05T17:38:12.5458120Z   Vulkan library is not found.  Executables may fail to start because the
2023-06-05T17:38:12.5562200Z   library may not be in rpath.  Define VULKAN_SDK environment or CMake
2023-06-05T17:38:12.5663960Z   variable to specify the SDK path.
```

iOS has no Vulkan and this error:

```
Neither VULKAN_SDK nor MOLTENVK_LIBRARY is defined. Vulkan backend will be disabled.
```

Or maybe this is closer to @rokups in topic...

**Comment by rbnpontes** [comment]
Sure.
I can't test iOS right now because i need to pay apple license to test on my devices, maybe next month will buy the license and bring your fixes

About VULKAN_SDK, this errors occurs because environment variables has not been configured correctly.

**Comment by eugeneko** [comment]
> I can't test iOS right now because i need to pay apple license to test on my devices, maybe next month will buy the license and bring your fixes

Maybe you shouldn't bother and we will just test it later via C# nuget, because C# deployment for Apple is easier.
It would be awesome even if we just test every platform except iOS. I can personally test Windows, UWP, Android, Web, so you don't need to bother with those if you don't have build tools.

And for Metal support in Mac/iOS GitHub Actions, you can just rely on GitHub and not pay anything heh.

PS: Known issue -- deferred rendering is broken because Diligent does not support read-only depth buffer. I will deal with it later.

**Comment by rokups** [comment]
If you have a Mac then you can at least try builds targeting simulator.
This goes a long way already

**Comment by rbnpontes** [comment]
Unfortunately i can't, build works but run build requires sign, and sign requires a license from apple

**Comment by eugeneko** [comment]
Texture3D and Texture2DArray are implemented. I also refactored texture implementation to reduce code size and improve flexibility.
